plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,183 @@
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = queryGroup.getName();
+        final Map<String, QueryGroup> previousGroups = metadata.queryGroups();
+
+        // check if maxQueryGroupCount will breach
+        if (previousGroups.size() + 1 > maxQueryGroupCount) {
+            logger.error("{} value exceeded its assigned limit of {}", QUERY_GROUP_COUNT_SETTING_NAME, maxQueryGroupCount);
+            throw new RuntimeException("Can't create more than " + maxQueryGroupCount + " QueryGroups in the system");
+        }
+
+        // check for duplicate name
+        Optional<QueryGroup> findExistingGroup = previousGroups.values()
+            .stream()
+            .filter(group -> group.getName().equals(groupName))
+            .findFirst();
+        if (findExistingGroup.isPresent()) {
+            logger.warn("QueryGroup with name {} already exists. Not creating a new one.", groupName);
+            throw new RuntimeException("QueryGroup with name " + groupName + " already exists. Not creating a new one.");
+        }
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            String resourceName = resourceType.getName();
+            double existingUsage = calculateExistingUsage(resourceName, previousGroups, groupName);
+            double newGroupUsage = getResourceLimitValue(resourceName, queryGroup.getResourceLimits());
+            if (existingUsage + newGroupUsage > 1) {
+                logger.error("Total resource allocation for {} will go above the max limit of 1.0", resourceName);
```

Comment:
Also, we should support percentage in addition to the ratio. @kaushalmahi12 - What do you think? I see this conversation is marked resolved - https://github.com/opensearch-project/OpenSearch/pull/13669/files#r1646712161